### PR TITLE
[playsuisse] Add new extractor

### PR DESCRIPTION
**Pull request by sbor23**
**This is a copy of the original [youtube-dl PR](https://github.com/ytdl-org/youtube-dl/pull/28142). Now supports subtitles!**

## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] [Searched](https://github.com/ytdl-org/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Read [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site)
- [x] Read [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) and adjusted the code to meet them
- [x] Covered the code with tests (note that PRs without tests will be REJECTED)
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Provide extractor for playsuisse.ch as requested in https://github.com/ytdl-org/youtube-dl/issues/27682 .
The website requires authentication, which can be provided to youtube-dl using the `--username,  --password` option, `--add-header` should also work. `.netrc` wasn't tested.

Account information for testing purposes:
E-Mail: tichis1964@einrot.com
Password: asdfasdf12341234!

### CODE REVIEW

File: yt_dlp/extractor/playsuisse.py
Comment (pukkandan) [inline yt_dlp/extractor/playsuisse.py:47]: if it has no effect, dont use it

File: yt_dlp/extractor/playsuisse.py
Comment (pukkandan) [inline yt_dlp/extractor/playsuisse.py:11]:
```suggestion
    _VALID_URL = r'https?://(?:www\.)?playsuisse\.ch/watch/(?P<id>[0-9]+)'
```

File: yt_dlp/extractor/playsuisse.py
Comment (pukkandan) [inline yt_dlp/extractor/playsuisse.py:63]:
```suggestion
        media_id = self._match_id(url)
```

File: yt_dlp/extractor/playsuisse.py
Comment (pukkandan) [inline yt_dlp/extractor/playsuisse.py:74]: title should just be the episode name imo. Make the other data available in the relevent keys like `series`, `episode` etc

File: yt_dlp/extractor/playsuisse.py
Comment (pukkandan) [inline yt_dlp/extractor/playsuisse.py:85]:
```suggestion
        thumbnails = [{
            'id': thumb.get('id'),
            'url': thumb['url']
        } for key, thumb in media_data.items() if thumb.get('url')]
```

File: yt_dlp/extractor/playsuisse.py
Comment (pukkandan) [inline yt_dlp/extractor/playsuisse.py:97]: wdym? The dash link gives 404? then why this code?

File: yt_dlp/extractor/playsuisse.py
Comment (pukkandan) [inline yt_dlp/extractor/playsuisse.py:104]:
```suggestion
        formats, subtitles = [], {}
        for media in media_data['medias']:
            if media.get('type') == 'HLS':
                f, subs = self._extract_m3u8_formats_and_subtitles(
                    media['url'], media_id, 'mp4', 'm3u8_native', m3u8_id='HLS', fatal=False)
            elif media.get('type') == 'DASH':
                f, subs = self._extract_mpd_formats_and_subtitles(
                    media['url'], media_id, mpd_id='dash', fatal=False))
            else:
                continue
            formats.append(f)
            self._merge_subtitles(subtitles, subs)
```

OR

```suggestion
        formats, subtitles = [], {}
        for media in media_data['medias']:
            if media.get('type') != 'HLS':
                continue
            f, subs = self._extract_m3u8_formats_and_subtitles(
                media['url'], media_id, 'mp4', 'm3u8_native', m3u8_id='HLS', fatal=False)
            formats.append(f)
            self._merge_subtitles(subtitles, subs)
```

depending on what the deal is with DASH

File: yt_dlp/extractor/playsuisse.py
Comment (pukkandan) [inline yt_dlp/extractor/playsuisse.py:115]: This looks copy pasted from the devtools and can probably be simplified

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/playsuisse.py
Comment (fstirlitz) [reply yt_dlp/extractor/playsuisse.py:104]:
```
self._merge_subtitles(subs, target=subtitles)
```

Comment (pukkandan) [reply yt_dlp/extractor/playsuisse.py:104]: in this case, `self._merge_subtitles(subtitles, subs)` and `self._merge_subtitles(subs, target=subtitles)` are the same, no?

Comment (fstirlitz) [reply yt_dlp/extractor/playsuisse.py:104]: https://github.com/yt-dlp/yt-dlp/blob/f1f6ca78b439343aa3f8ef44f803befd682a3d37/yt_dlp/extractor/common.py#L3483-L3491

File: yt_dlp/extractor/playsuisse.py
Comment (sbor23) [reply yt_dlp/extractor/playsuisse.py:115]: Thx for your review, will have a look shortly.

File: yt_dlp/extractor/playsuisse.py
Comment (sbor23) [reply yt_dlp/extractor/playsuisse.py:47]: The point is it _should_ have an effect. The platform offers multiplexes multiple audio streams to the same media id. In the browser this is switchable (sets the header), but it would work in ytdl. I debugged into the request call and didn't see the headers stripped anywhere...

**Comment by pukkandan** [comment]
Please don't force push. It makes reviewing harder

File: yt_dlp/extractor/playsuisse.py
Comment (pukkandan) [inline yt_dlp/extractor/playsuisse.py:86]: unnecessary code - maybe u forgot to remove this?

File: yt_dlp/extractor/playsuisse.py
Comment (pukkandan) [inline yt_dlp/extractor/playsuisse.py:55]:
```suggestion
        response = self._download_json(
            'https://4bbepzm4ef.execute-api.eu-central-1.amazonaws.com/prod/graphql',
            media_id, data=json.dumps({
                'operationName': 'AssetWatch',
                'query': self._GRAPHQL_QUERY,
                'variables': {'assetId': media_id}
            }).encode('utf-8'), headers={'Content-Type': 'application/json'})
```

File: yt_dlp/extractor/playsuisse.py
Comment (pukkandan) [inline yt_dlp/extractor/playsuisse.py:9]:
```suggestion
```

File: yt_dlp/extractor/playsuisse.py
Comment (pukkandan) [inline yt_dlp/extractor/playsuisse.py:102]: add test for these

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/playsuisse.py
Comment (pukkandan) [reply yt_dlp/extractor/playsuisse.py]: I meant remove the unused code. You can keep the `TODO` comment if u think this is something that should be looked into (but explain it better)

File: yt_dlp/extractor/playsuisse.py
Comment (sbor23) [reply yt_dlp/extractor/playsuisse.py:99]: done

File: yt_dlp/extractor/playsuisse.py
Comment (pukkandan) [reply yt_dlp/extractor/playsuisse.py:106]: you didn't address this

Comment (sbor23) [reply yt_dlp/extractor/playsuisse.py:106]: I did, it was around 200 lines more. I kept only the stuff that is actually used or could potentially be mapped to yt-dlp metadata fields and actually holds useful info. I wouldn't want to cut it down more

Comment (pukkandan) [reply yt_dlp/extractor/playsuisse.py:106]: ah, I missed the change coz of porce-push

**Comment by sbor23** [comment]
@pukkandan is this PR ready or am I missing something?

**Comment by paavo** [comment]
Any eta when this will be reviewed @pukkandan?

File: yt_dlp/extractor/playsuisse.py
Comment (pukkandan) [reply yt_dlp/extractor/playsuisse.py:106]: still lots of unnecessary fields

**Comment by pukkandan** [comment]
I have made some untested changed (due to geo-blocking). test it and let me know

**Comment by sbor23** [comment]
> I have made some untested changed (due to geo-blocking). test it and let me know

@pukkandan Looks good now!

File: yt_dlp/extractor/playsuisse.py
Comment (pukkandan) [inline yt_dlp/extractor/playsuisse.py:98]:
```suggestion
        thumbnails = [{
            'id': thumb.get('id'),
            'url': thumb['url']
        } for thumb in traverse_obj(media_data, lambda k, v: k.startswith('thumbnail') and v['url'])]
```

why did you revert this!?

Comment (sbor23) [reply yt_dlp/extractor/playsuisse.py:98]: Because I tested it and this part failed. It returns `None` somewhere, it's not like this check wasn't there for a reason...

Comment (pukkandan) [reply yt_dlp/extractor/playsuisse.py:98]: the check is implied in `traverse_obj`.

test with this suggestion after merging with master (there has been some change to traverse_obj since the the PR) and if there is an issue, let me know exactly what it is

File: yt_dlp/extractor/playsuisse.py
Comment (goggle) [inline yt_dlp/extractor/playsuisse.py:46]:
```suggestion
_NETRC_MACHINE = 'playsuisse'

```
We might want to add this to make it work with netrc.

File: yt_dlp/extractor/playsuisse.py
Comment (goggle) [inline yt_dlp/extractor/playsuisse.py:8]:
```suggestion
class PlaySuisseIE(InfoExtractor):
    IE_DESC = 'playsuisse.ch'
```
This PR doesn't work for me without this change. Otherwise I get this error:
```
ERROR: 'PlaySuisseIE' object has no attribute 'IE_DESC'
```
I'm not sure if that's the right approach though (@pukkandan ?)

Comment (pukkandan) [reply yt_dlp/extractor/playsuisse.py:8]: You must be applying the PR on a commit before https://github.com/yt-dlp/yt-dlp/commit/231025c4632d7a48540b18997c715a8df8a6ff8e

File: yt_dlp/extractor/playsuisse.py
Comment (pukkandan) [reply yt_dlp/extractor/playsuisse.py:46]: The extractor hasn't added login support. So _NETRC_MACHINE should not be added

Comment (goggle) [reply yt_dlp/extractor/playsuisse.py:46]: You're right, I've only read the description...

**Comment by thde** [comment]
Apperantly there is now a V2 for the GraphQL API (V1 is still working but deprecated):

```diff
--- a/yt_dlp/extractor/playsuisse.py
+++ b/yt_dlp/extractor/playsuisse.py
@@ -46,7 +46,7 @@ class PlaySuisseIE(InfoExtractor):

     _GRAPHQL_QUERY = '''
         query AssetWatch($assetId: ID!) {
-            asset(assetId: $assetId) {
+            assetV2(id: $assetId) {
                 name
                 description
                 duration
@@ -69,7 +69,7 @@ class PlaySuisseIE(InfoExtractor):
                 }
             }
         }
-        fragment ImageDetails on Image {
+        fragment ImageDetails on AssetImage {
             id
             url
         }'''
@@ -86,7 +86,7 @@ def _get_media_data(self, media_id):
             }).encode('utf-8'),
             headers={'Content-Type': 'application/json', 'locale': 'de'})

-        return response['data']['asset']
+        return response['data']['assetV2']

     def _real_extract(self, url):
         media_id = self._match_id(url)
```

**Comment by sbor23** [comment]
@thde Thanks for the heads-up, I will give the changes a try soon. Unfortunately I'm also geoblocked at the moment due to travel, hence need a couple of weeks to get it tested.

**Comment by thde** [comment]
Thanks @sbor23! I actually found out something else: With shows ([this](https://www.playsuisse.ch/detail/1115687/search-montreux-de/montreux?locale=de) at least) the download doesn't work, because the ID is pointing to a show with multiple episodes:
```
https://www.playsuisse.ch/watch/1115687/51/search-montreux-de?episodeId=1112663&locale=de
```
The actual ID of the video is the `episodeId` GET param in this case.

One could just use the `episodeId` of course. But with a small change to the query, all episodes can be looked up and it could be handeled as a playlist I think:

```diff
--- a/yt_dlp/extractor/playsuisse.py
+++ b/yt_dlp/extractor/playsuisse.py
@@ -47,9 +47,18 @@ class PlaySuisseIE(InfoExtractor):
     _GRAPHQL_QUERY = '''
         query AssetWatch($assetId: ID!) {
             assetV2(id: $assetId) {
+                ...Asset
+                episodes {
+                    ...Asset
+                }
+            }
+        }
+        fragment Asset on AssetV2 {
                 name
                 description
                 duration
+                seriesName
+                seasonNumber
                 episodeNumber
                 medias {
                     type
@@ -67,7 +76,6 @@ class PlaySuisseIE(InfoExtractor):
                 thumbnail2x3WithTitle {
                     ...ImageDetails
                 }
-            }
         }
         fragment ImageDetails on AssetImage {
             id
```

**Comment by sbor23** [comment]
@pukkandan The remaining points where addressed. IMO this is good to merge, let me know if I should squash.

File: yt_dlp/extractor/playsuisse.py
Comment (pukkandan) [inline yt_dlp/extractor/playsuisse.py:1]:
```suggestion
```

File: yt_dlp/extractor/playsuisse.py
Comment (pukkandan) [inline yt_dlp/extractor/playsuisse.py:119]:
```suggestion
            info.update({
                '_type': 'playlist',
                'entries': (self._extract_single(episode) for episode in media_data['episodes'])
            })
```

File: yt_dlp/extractor/playsuisse.py
Comment (pukkandan) [inline yt_dlp/extractor/playsuisse.py:130]:
```suggestion
                media['url'], media_data['id'], 'mp4', m3u8_id='HLS', fatal=False)
```

**Review by pukkandan — CHANGES_REQUESTED** [review]

**Pull request merged**
